src/ioctls/mod.rs

```diff
@@ -26,6 +29,100 @@ pub mod vm;
+    pub(crate) fn mmap_from_fd<F: AsRawFd>(fd: &F) -> Result<Self> {
+        // SAFETY: We trust the sysconf libc function and we're calling it
+        // with a correct parameter.
+        let page_size = match unsafe { libc::sysconf(libc::_SC_PAGESIZE) } {
+            -1 => return Err(errno::Error::last()),
+            ps => ps as usize,
+        };
+
+        let offset = KVM_COALESCED_MMIO_PAGE_OFFSET * page_size as u32;
+        // SAFETY: KVM guarantees that there is a page at offset
+        // KVM_COALESCED_MMIO_PAGE_OFFSET * PAGE_SIZE if the appropriate
+        // capability is available. If it is not, the call will simply
+        // fail.
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                page_size,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset.into(),
+            )
+        };
+        if addr == libc::MAP_FAILED {
+            return Err(errno::Error::last());
+        }
+        Ok(Self {
+            addr: addr.cast::<kvm_coalesced_mmio_ring>(),
+            page_size,
+        })
+    }
+
+    /// Compute the size of the MMIO ring.
+    /// Taken from [include/uapi/linux/kvm.h](https://elixir.bootlin.com/linux/v6.6/source/include/uapi/linux/kvm.h#L562)
+    const fn ring_max(&self) -> usize {
+        (self.page_size - size_of::<kvm_coalesced_mmio_ring>()) / size_of::<kvm_coalesced_mmio>()
+    }
+
+    /// Gets a mutable reference to the ring
+    fn ring_mut(&mut self) -> &mut kvm_coalesced_mmio_ring {
+        // SAFETY: We have a `&mut self` and the pointer is private, so this
+        // access is exclusive.
+        unsafe { &mut *self.addr }
+    }
+
+    /// Reads a single entry from the MMIO ring.
+    ///
+    /// # Returns
+    ///
+    /// An entry from the MMIO ring buffer, or [`None`] if the ring is empty.
+    pub(crate) fn read_entry(&mut self) -> Option<kvm_coalesced_mmio> {
+        let ring_max = self.ring_max();
+
+        let ring = self.ring_mut();
+        if ring.first == ring.last {
+            return None;
+        }
+
+        let entries = ring.coalesced_mmio.as_ptr();
+        // SAFETY: `ring.first` is an `u32` coming from mapped memory filled
+        // by the kernel, so we trust it. `entries` is a pointer coming from
+        // mmap(), so pointer arithmetic cannot overflow. We have a `&mut self`,
+        // so nobody else has access to the contents of the pointer.
+        let elem = unsafe { entries.add(ring.first as usize).read() };
+        ring.first = (ring.first + 1) % ring_max as u32;
+
+        Some(elem)
+    }
+}
+
+impl Drop for KvmCoalescedIoRing {
+    fn drop(&mut self) {
+        // SAFETY: This is safe because we mmap the page ourselves, and nobody
+        // else is holding a reference to it.
+        unsafe {
+            libc::munmap(self.addr.cast::<libc::c_void>(), self.page_size);
```

Comment:
Nit: I think it can infer the type of this cast.
```suggestion
            addr: addr.cast(),
```